File: src/python/multithreading/10_schedule_every_n_sec.py
"""
A simple program to demonstrate a scheduler that runs a function every n seconds.
The scheduler runs the function in a separate thread and can be stopped after a specified duration.
"""

import time
from threading import Thread


class Scheduler:
    """
    A scheduler class that repeatedly executes a given function every n seconds.
    """

    def __init__(self, n, function, *args, **kwargs):
        """
        Initialize the scheduler with the interval n, the function to execute, and any arguments or keyword arguments.

        :param n: The interval in seconds between executions of the function.
        :param function: The function to execute.
        :param args: Positional arguments for the function.
        :param kwargs: Keyword arguments for the function.
        """

        def target():
            while self.flag:
                function(*args, **kwargs)
                time.sleep(n)

        self.flag = True
        self.thread = Thread(target=target)
        self.thread.start()

    def stop(self):
        """
        Stop scheduling new function calls.
        If the previous function call has not yet completed, it will not be terminated.
        """
        self.flag = False
        self.thread.join()


def sample_function(message):
    print(message)


if __name__ == "__main__":
    scheduler = Scheduler(1, sample_function, "It works!")

    # Wait 10 seconds and stop the scheduler
    time.sleep(10)
    scheduler.stop()

File: src/cpp/multithreading/10_schedule_every_n_sec.cpp
#include <iostream>
#include <thread>
#include <atomic>
#include <chrono>
#include <functional>

class Scheduler {
public:
    Scheduler(int n, std::function<void()> function)
        : flag(true), interval(n), task(std::move(function)) {
        scheduler_thread = std::thread([this] {
            while (flag) {
                task();
                std::this_thread::sleep_for(std::chrono::seconds(interval));
            }
        });
    }

    ~Scheduler() {
        stop();
    }

    void stop() {
        flag = false;
        if (scheduler_thread.joinable()) {
            scheduler_thread.join();
        }
    }

private:
    std::atomic<bool> flag;
    int interval;
    std::function<void()> task;
    std::thread scheduler_thread;
};

void print_message() {
    std::cout << "It works" << std::endl;
}

int main() {
    Scheduler scheduler(1, print_message);

    // wait 10 seconds and stop the scheduler
    std::this_thread::sleep_for(std::chrono::seconds(10));
    scheduler.stop();

    return 0;
}
